presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduSplitManager.java

```diff
@@ -59,4 +76,72 @@ public ConnectorSplitSource getSplits(
                 throw new IllegalArgumentException("Unknown splitSchedulingStrategy: " + splitSchedulingStrategy);
         }
     }
+
+    private class KuduDynamicFilteringSplitSource
+            implements ConnectorSplitSource
+    {
+        private final CompletableFuture<ConnectorSplitSource> future;
+
+        private KuduDynamicFilteringSplitSource(
+                ConnectorTableHandle table,
+                SplitSchedulingStrategy splitSchedulingStrategy,
+                DynamicFilter dynamicFilter,
+                long blockingTimeoutMillis)
+        {
+            this.future = whenCompleted(dynamicFilter)
+                    .completeOnTimeout(null, blockingTimeoutMillis, MILLISECONDS)
+                    .thenApply(o -> getSplitSource(table, splitSchedulingStrategy, dynamicFilter));
```

Comment:
extract `new KuduDynamicFilteringSplitSource(table, splitSchedulingStrategy, dynamicFilter, blockingTimeoutMillis);` to internal method and don't call top-level `getSplitSource` (it creases tricky recursive flow)

Reply:
updated